dev/kyuubi-extension-spark-3-1/src/main/scala/org/apache/kyuubi/sql/watchdog/MaxPartitionStrategy.scala

```diff
@@ -0,0 +1,176 @@
+ * 1 Check if scan exceed maxPartition
+ * 2 Check if Using partitionFilter on partitioned table
+ * This Strategy Add Planner Strategy after LogicalOptimizer
+ */
+case class MaxPartitionStrategy(session: SparkSession)
+  extends Strategy
+  with SQLConfHelper
+  with PruneFileSourcePartitionHelper {
+  override def apply(plan: LogicalPlan): Seq[SparkPlan] = {
+    conf.getConf(KyuubiSQLConf.WATCHDOG_MAX_PARTITIONS) match {
+      case Some(maxScanPartitions) => plan match {
+          case ScanOperation(_, _, relation: HiveTableRelation) if relation.isPartitioned =>
+            relation.prunedPartitions match {
+              case Some(prunedPartitions) =>
+                if (prunedPartitions.size > maxScanPartitions) {
+                  throw new MaxPartitionExceedException(
+                    s"""
+                       |SQL job scan hive partition: ${prunedPartitions.size}
+                       |exceed restrict of hive scan maxPartition $maxScanPartitions
+                       |You should optimize your SQL logical according partition structure
+                       |or shorten query scope such as p_date, detail as below:
+                       |Table: ${relation.tableMeta.qualifiedName}
+                       |Owner: ${relation.tableMeta.owner}
+                       |Partition Structure: ${relation.partitionCols.map(_.name).mkString(" -> ")}
+                       |""".stripMargin)
+                } else {
+                  Nil
+                }
+              case _ =>
+                val totalPartitions = session
+                  .sessionState.catalog.externalCatalog.listPartitionNames(
+                    relation.tableMeta.database,
+                    relation.tableMeta.identifier.table)
+                if (totalPartitions.size > maxScanPartitions) {
+                  throw new MaxPartitionExceedException(
+                    s"""
+                       |Your SQL job scan a whole huge table without any partition filter,
+                       |You should optimize your SQL logical according partition structure
+                       |or shorten query scope such as p_date, detail as below:
+                       |Table: ${relation.tableMeta.qualifiedName}
+                       |Owner: ${relation.tableMeta.owner}
+                       |Partition Structure: ${relation.partitionCols.map(_.name).mkString(" -> ")}
+                       |""".stripMargin)
+                } else {
+                  Nil
+                }
+            }
+          case ScanOperation(
+                _,
+                filters,
+                relation @ LogicalRelation(
+                  fsRelation @ HadoopFsRelation(
+                    fileIndex: InMemoryFileIndex,
+                    partitionSchema,
+                    _,
+                    _,
+                    _,
+                    _),
+                  _,
+                  _,
+                  _)) if fsRelation.partitionSchemaOption.isDefined =>
```

Comment:
actually, I really don't like such code format ... Is there a way to relax the new line style ? cc @pan3793 @zhouyifan279

Reply:
this is not good..

Reply:
kindly ping @zhouyifan279 @pan3793